### PR TITLE
PY: python bindings

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What
Adding python-bindings for the following UCP functionality:
1. Init (using context creation, and worker creation); Finalize
2. Endpoint creation (using listen-connect-accept mode)
3. Send/Recv (using tag_{send/recv}_nb + worker_progress)
4. Added ability to operate in python programs that use coroutines

## Why ?
Internal request to enable python-based frameworks which wanted to leverage UCX capabilities (including GPU support)

## How ?
Using Cython and C wrapper functions.

cc @mrocklin @TomAugspurger @shamisp @yosefe

### CODE REVIEW

File: bindings/python/tests/single.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2019, Anaconda, Inc. All rights reserved.
```
Comment (shamisp) [inline bindings/python/tests/single.py]: Is this 3rd party copyright ? what was the original license ?

File: bindings/python/Makefile [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2018-2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (shamisp) [inline bindings/python/Makefile]: Can we use standard autotools to generate the makefile ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8654/ for details (Mellanox internal link).

File: bindings/python/src/ucp_py.pyx [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+        """Blind receive operation"""
+
+        recv_msg = ucp_msg(None)
+        recv_msg.ucp_ep = self.ucp_ep
+        recv_future = CommFuture(recv_msg)
+        ucp_py_ep_post_probe()
+        return recv_future
+
+    def recv(self, ucp_msg msg, len):
+        """Receive operation of length `len`
```
Comment (noaos) [inline bindings/python/src/ucp_py.pyx]: Please stick with Python's docstring conventional structure throughout the code.

File: bindings/python/src/ucp_py.pyx [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+
+        Returns
+        -------
+        CommFuture object
+        """
+
+        msg.ctx_ptr = ucp_py_recv_nb(self.ucp_ep, msg.buf, len)
```
Comment (noaos) [inline bindings/python/src/ucp_py.pyx]: I see you avoid checking arguments or try/excepts. Wouldn't it be better to properly report errors to users?

File: bindings/python/Makefile [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+pybind/libucp_py_ucp_fxns.a: pybind/ucp_py_ucp_fxns.o pybind/buffer_ops.o
+	ar rcs $@ $^
+
+pybind/%.o: pybind/%.c
+	$(CC) -shared -fPIC $(CPPFLAGS) -c $^ -o $@ $(LDFLAGS)
+
+install: pybind/libucp_py_ucp_fxns.a
+	cd pybind && \
+	python3 setup.py build_ext && \
```
Comment (yosefe) [inline bindings/python/Makefile]: need to use autotools rather than assuming paths of cuda, python, gcc ...

File: bindings/python/Makefile [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+	python3 -m pip install -e .
```
Comment (yosefe) [inline bindings/python/Makefile]: does it work with rpm packaging?

**Comment by Akshay-Venkatesh** [comment]
Using this for external ref. Reopening until new pr is submitted

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9995/ for details (Mellanox internal link).

**Comment by noaos** [comment]
> 
> 
> Can one of the admins verify this patch?

Not all the comments were addressed.

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>ABORTED</b> on 1 workers (click for details)</summary>

**Note:** the logs will be deleted after 15-Nov-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/15f07d2389935922ea32a568dedc7aeb/raw/424d665177182551fb0fd3ad8406eeeea62871df/_main) | :question: ABORTED

**Pull request closed without merging**